serveradmin/serverdb/query_committer.py

```diff
@@ -616,6 +626,91 @@ def _validate_real_attributes(servertype, real_attributes):     # NOQA: C901
+        return changed
+
+    # Forbid deleting the servertype attribute
+    for change in servertype_changes:
+        if (
+            change['servertype']['action'] != 'update' or
+            'new' not in change['servertype'] or
+            not change['servertype']['new']
+        ):
+            CommitError('Cannot delete servertype')
+
+    # Get attributes of the new and old servertypes to check what changes
+    involved_servertypes = chain(*[
+        [change['servertype']['old'], change['servertype']['new']]
+        for change in servertype_changes
+    ])
+
+    try:
+        servertype_attributes = _get_servertype_attributes(
+            servertypes=involved_servertypes
+        )
+    except Servertype.DoesNotExist:
+        raise CommitError('Unknown servertype')
+
+    # The fetching and materializing of servers here is duplicating queries
+    # done in the commit_query transaction. We need to do this to resolve all
+    # the real attribute changes, resulting from a servertype change, before
+    # sending the pre_commit signal. Alternatively we could send the
+    # pre_commit signal inside the commit_query transaction but sending the
+    # signal can take many seconds if a receiver misbehaves.
+    with transaction.atomic():
+        changed_servers = _fetch_servers(
+            set(c['object_id'] for c in servertype_changes)
+        )
+        changed_objects = _materialize(changed_servers, joined_attributes)
+
+
+    # Mark all attributes, not available on the new servertype, for deletion.
+    for change in changed:
+        if change not in servertype_changes:
+            continue
+
+        old = set(servertype_attributes[change['servertype']['old']].keys())
+        new = set(servertype_attributes[change['servertype']['new']].keys())
+        for del_attr in old - new:
+            if del_attr in change:
+                raise CommitError(
+                    'Attribute value of {} needs to be deleted for servertype '
+                    'change, but was modified in this commit.'.format(del_attr)
+                )
+
+            changed_object = changed_objects[change['object_id']]
+            if changed_object[del_attr]:
+                change[del_attr] = {
+                    'action': 'update',
+                    'old': changed_object[del_attr],
+                    'new': None
+                }
+
+    # TODO: Handle attributes beeing required on the new servertype that are
+    # not available on the old servertype.  Changing servertypes means the
+    # available attributes change.  Adminapi and servershell are not equipped
+    # for this.  Hence we can not set values for these attributes in the same
+    # commit for now.  While this is not super clean, the user will be forced
+    # to set a correct value once he opens an edit dialog for this object in
+    # servershell.  I think that's good enough for now as changing servertypes
+    # is an edgecase anyway.
+
+    return None
```

Comment:
No need

Reply:
Fair enough, though mypy recommends this.

Reply:
Does it still after you change the previous `return`?

Reply:
Honestly I currently don't even have it installed. I just made it a habit as mypy by default recommends to explicitly return everywhere, which I adapted. If I had it installed I would have caught the other return though ;) I'll just remove it for now and revisit once I get around to putting type hints in here.